src/algebra/category/Semigroup/basic.lean

```diff
@@ -0,0 +1,217 @@
+
+/-- Construct a bundled `Magma` from the underlying type and typeclass. -/
+@[to_additive]
+def of (M : Type u) [has_mul M] : Magma := bundled.of M
+
+/-- Construct a bundled `Magma` from the underlying type and typeclass. -/
+add_decl_doc AddMagma.of
+
+@[to_additive]
+instance : inhabited Magma := ⟨Magma.of empty⟩
+
+@[to_additive]
+instance (M : Magma) : has_mul M := M.str
+
+@[simp, to_additive] lemma coe_of (R : Type u) [has_mul R] : (Magma.of R : Type u) = R := rfl
+
+end Magma
+
+/-- The category of semigroups and semigroup morphisms. -/
+@[to_additive AddSemigroup]
+def Semigroup : Type (u+1) := bundled semigroup
+
+/-- The category of additive semigroups and semigroup morphisms. -/
+add_decl_doc AddSemigroup
+
+namespace Semigroup
+
+@[to_additive]
+instance : bundled_hom.parent_projection semigroup.to_has_mul := ⟨⟩
+
+attribute [derive [has_coe_to_sort, large_category, concrete_category]] Semigroup AddSemigroup
+
+/-- Construct a bundled `Semigroup` from the underlying type and typeclass. -/
+@[to_additive]
+def of (M : Type u) [semigroup M] : Semigroup := bundled.of M
+
+/-- Construct a bundled `AddSemigroup` from the underlying type and typeclass. -/
+add_decl_doc AddSemigroup.of
+
+@[to_additive]
+instance : inhabited Semigroup := ⟨Semigroup.of empty⟩
+
+@[to_additive]
+instance (M : Semigroup) : semigroup M := M.str
+
+@[simp, to_additive] lemma coe_of (R : Type u) [semigroup R] : (Semigroup.of R : Type u) = R := rfl
+
+@[to_additive has_forget_to_AddMagma]
+instance has_forget_to_Magma : has_forget₂ Semigroup Magma := bundled_hom.forget₂ _ _
+
+end Semigroup
+
+-- We verify that the coercions of morphisms to functions work correctly:
+example {R S : Magma}     (f : R ⟶ S) : (R : Type) → (S : Type) := f
+example {R S : Semigroup} (f : R ⟶ S) : (R : Type) → (S : Type) := f
+
+-- We verify that when constructing a morphism in `Semigroup`,
+-- when we construct the `to_fun` field, the types are presented as `↥R`,
+-- rather than `R.α` or (as we used to have) `↥(bundled.map semigroup.to_has_mul R)`.
+example (R : Semigroup.{u}) : R ⟶ R :=
+{ to_fun := λ X,
+  begin
+    match_target (R : Type u),
+    exact X
+  end ,
+  map_mul' := λ x y,
+  rfl, }
+
+
+variables {X Y : Type u}
+
+section
+variables [has_mul X] [has_mul Y]
+
+/-- Build an isomorphism in the category `Magma` from a `mul_equiv` between `has_mul`s. -/
+@[to_additive add_equiv.to_AddMagma_iso "Build an isomorphism in the category `AddMagma` from
+an `add_equiv` between `has_add`s."]
+def mul_equiv.to_Magma_iso (e : X ≃* Y) : Magma.of X ≅ Magma.of Y :=
+{ hom := e.to_mul_hom,
+  inv := e.symm.to_mul_hom }
+
+@[simp, to_additive add_equiv.to_AddMagma_iso_hom]
+lemma mul_equiv.to_Magma_iso_hom {e : X ≃* Y} : e.to_Magma_iso.hom = e.to_mul_hom := rfl
+@[simp, to_additive add_equiv.to_AddMagma_iso_inv]
+lemma mul_equiv.to_Magma_iso_inv {e : X ≃* Y} : e.to_Magma_iso.inv = e.symm.to_mul_hom := rfl
```

Comment:
Can these lemmas be constructed merely by putting `@[simps]` on `mul_equiv.to_Magma_iso`?

Reply:
Yes, therefore I also changed it in the Mon file to work with `@[simps]`.